main.tf

```
@@ -335,6 +336,23 @@ data "aws_iam_policy_document" "bucket_policy" {
       }
     }
   }
+
+  dynamic "statement" {
+    for_each = var.privileged_principal_arns
+
+    content {
+      sid     = "AllowPrivilegedPrincipal[${ { for k, v in keys(var.privileged_principal_arns) : v => k }[statement.key]}]" # add indices to Sid
+      actions = var.privileged_principal_actions
+      resources = distinct(flatten([
+        "arn:${data.aws_partition.current.partition}:s3:::${join("", aws_s3_bucket.default.*.id)}",
+        formatlist("arn:${data.aws_partition.current.partition}:s3:::${join("", aws_s3_bucket.default.*.id)}/%s*", statement.value),
+      ]))
+      principals {
+        type        = "AWS"
+        identifiers = [statement.key]
```

Clever, but I do not like regenerating the list on every iteration
```suggestion
    for_each = keys(var.privileged_principal_arns)

    content {
      sid     = "AllowPrivilegedPrincipal[${statement.key}]" # add indices to Sid
      actions = var.privileged_principal_actions
      resources = distinct(flatten([
        "arn:${data.aws_partition.current.partition}:s3:::${join("", aws_s3_bucket.default.*.id)}",
        formatlist("arn:${data.aws_partition.current.partition}:s3:::${join("", aws_s3_bucket.default.*.id)}/%s*", var.privileged_principal_arns[statement.value]),
      ]))
      principals {
        type        = "AWS"
        identifiers = [statement.value]
```